### PR TITLE
THREESCALE-8502 3scale operator to have an ability to deal with multiple release paths

### PR DESCRIPTION
**Pull request by eguzki**
### what

[THREESCALE-8502](https://issues.redhat.com/browse/THREESCALE-8502) 3scale operator to have an ability to deal with multiple release paths

### how

Upgrade procedure in reconciliation loop. Actually, most of it was already in the normal reconciliation loop. Only DC triggers and pod template labels were moved

custom release upgrade procedure has been removed.

### verification steps

Deploy 2.12

```
git checkout 3scale-2.12.0-GA
```
```
make run
```

```yaml
k apply -f - <<EOF
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: null
  name: aws-auth
stringData:
  AWS_ACCESS_KEY_ID: testID
  AWS_SECRET_ACCESS_KEY: testkey
  AWS_BUCKET: testbucket
  AWS_REGION: us-east-1
type: Opaque
EOF
```
```yaml
k apply -f - <<EOF
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager1
spec:
  wildcardDomain: eguzki.apps.dev-eng-ocp4-operator.dev.3sca.net
  resourceRequirementsEnabled: false
  system:
    fileStorage:
      simpleStorageService:
        configurationSecretRef:
          name: aws-auth
EOF
```
wait for deployment to be ready
```
oc wait --for=condition=available apimanager/apimanager1 --timeout=-1s
apimanager.apps.3scale.net/apimanager1 condition met
```
Check apimanager version
```
k get apimanager apimanager1 -o yaml | yq e '.metadata.annotations' -
apps.3scale.net/apimanager-threescale-version: "2.12"
apps.3scale.net/threescale-operator-version: 0.9.0
```
Check imagestreams, for instance amp-apicast
```yaml
k get is amp-apicast -o yaml | yq e '.spec.tags' -
- annotations:
    openshift.io/display-name: AMP APIcast 2.12
  from:
    kind: DockerImage
    name: quay.io/3scale/3scale212:apicast-3scale-2.12.0-GA
  generation: 2
  importPolicy: {}
  name: "2.12"
  referencePolicy:
    type: Source
```
Check deployment config's triggers, for instance, apicast-production
```yaml
k get dc apicast-production -o yaml | yq e '.spec.triggers' -
- type: ConfigChange
- imageChangeParams:
    automatic: true
    containerNames:
      - system-master-svc
      - apicast-production
    from:
      kind: ImageStreamTag
      name: amp-apicast:2.12
      namespace: eguzki
    lastTriggeredImage: quay.io/3scale/3scale212@sha256:f19731b5fe9bf4ba60977d1498eff5082836a152662d7fe310928f10e7e5658a
  type: ImageChange
```

Kill 2.12 operator with `CTRL+c` and move git workspace to `apimanager-upgrade-in-main-reconciler-loop` branch

```
git checkout apimanager-upgrade-in-main-reconciler-loop
```

Run the controller to perform upgrade

```
make run
```
wait for deployment to be ready
```
oc wait --for=condition=available apimanager/apimanager1 --timeout=-1s
apimanager.apps.3scale.net/apimanager1 condition met
```
Check apimanager version has been updated to 2.13
```
k get apimanager apimanager1 -o yaml | yq e '.metadata.annotations' -
apps.3scale.net/apimanager-threescale-version: "2.13"
apps.3scale.net/threescale-operator-version: 0.10.0
```

Check that imagestream objects have a new tag for the 2.13

```yaml
k get is amp-apicast -o yaml | yq e '.spec.tags' -
- annotations:
    openshift.io/display-name: AMP APIcast 2.12
  from:
    kind: DockerImage
    name: quay.io/3scale/3scale212:apicast-3scale-2.12.0-GA
  generation: 2
  importPolicy: {}
  name: "2.12"
  referencePolicy:
    type: Source
- annotations:
    openshift.io/display-name: AMP APIcast 2.13
  from:
    kind: DockerImage
    name: quay.io/3scale/apicast:latest
  generation: 4
  importPolicy: {}
  name: "2.13"
  referencePolicy:
    type: Source
```

Check that deployment config triggers has reference updated to 2.13

```yaml
- imageChangeParams:
    automatic: true
    containerNames:
      - system-master-svc
      - apicast-production
    from:
      kind: ImageStreamTag
      name: amp-apicast:2.13
      namespace: eguzki
    lastTriggeredImage: quay.io/3scale/apicast@sha256:31c9d27750f0ddc3e08269058ac07e0d490b38f25c1ab6f36431f93d99acb9ed
  type: ImageChange
```

### CODE REVIEW

File: pkg/reconcilers/deploymentconfig.go [hunk 1]
```diff
@@ -151,3 +159,53 @@ func DeploymentConfigEnvVarReconciler(desired, existing *appsv1.DeploymentConfig
...
 }
+
+func findDeploymentTriggerOnImageChange(triggerPolicies []appsv1.DeploymentTriggerPolicy) (int, error) {
+	result := -1
+	for i := range triggerPolicies {
+		if triggerPolicies[i].Type == appsv1.DeploymentTriggerOnImageChange {
+			if result != -1 {
+				return -1, fmt.Errorf("found more than one imageChangeParams Deployment trigger policy")
+			}
+			result = i
```
Comment (KevFan) [inline pkg/reconcilers/deploymentconfig.go:170]:
```suggestion
			result = i
			break
```
Minor, should there be a `break` here once the correct DeploymentTrigger is found ? :thinking:

**Comment by KevFan** [comment]
Verified ! :+1:  Just have one minor comment

File: pkg/reconcilers/deploymentconfig.go (hunk 1 above)
Comment (eguzki) [reply pkg/reconcilers/deploymentconfig.go:170]: good. Adding the break disables detecting multiple triggers on image change, so I will remove the error above as well.

Comment (eguzki) [reply pkg/reconcilers/deploymentconfig.go:168]: done

**Review by KevFan — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**